File: sdk/python/genvis_client.py
import os
import time
from typing import Any, Dict, Optional

import requests
from openai import OpenAI


def get_config() -> Dict[str, str]:
    api_key = os.getenv("GENVIS_API_KEY")
    base_url = os.getenv("GENVIS_BASE_URL", "https://genvis.xyz/v1")

    if not api_key or api_key == "sk-your-api-key":
        raise RuntimeError("Set GENVIS_API_KEY in your environment or .env file.")

    return {"api_key": api_key, "base_url": base_url.rstrip("/")}


def create_openai_client() -> OpenAI:
    config = get_config()
    return OpenAI(api_key=config["api_key"], base_url=config["base_url"])


def submit_json_video_task(payload: Dict[str, Any]) -> Dict[str, Any]:
    config = get_config()
    response = requests.post(
        f"{config['base_url']}/video/generations",
        headers={
            "Authorization": f"Bearer {config['api_key']}",
            "Content-Type": "application/json",
        },
        json=payload,
        timeout=60,
    )
    return _read_json_response(response)


def get_json_video_task(task_id: str) -> Dict[str, Any]:
    config = get_config()
    response = requests.get(
        f"{config['base_url']}/video/generations/{task_id}",
        headers={"Authorization": f"Bearer {config['api_key']}"},
        timeout=60,
    )
    return _read_json_response(response)


def poll_video_task(
    task_id: str,
    interval_seconds: int = 5,
    timeout_seconds: int = 600,
) -> Dict[str, Any]:
    started_at = time.time()

    while time.time() - started_at < timeout_seconds:
        task = get_json_video_task(task_id)
        status = task.get("status") or task.get("data", {}).get("status")

        if status in {"completed", "succeeded", "success"}:
            return task

        if status in {"failed", "cancelled", "canceled"}:
            raise RuntimeError(f"Video task failed: {task}")

        time.sleep(interval_seconds)

    raise TimeoutError(f"Video task timed out after {timeout_seconds} seconds: {task_id}")


def _read_json_response(response: requests.Response) -> Dict[str, Any]:
    try:
        data: Dict[str, Any] = response.json()
    except ValueError:
        data = {"raw": response.text}

    if not response.ok:
        error = data.get("error") or {}
        message = error.get("message") or data.get("message") or response.reason
        raise RuntimeError(f"Genvis API error {response.status_code}: {message}")

    return data

File: sdk/node/genvis-client.mjs
import OpenAI from 'openai';

export function getConfig() {
  const apiKey = process.env.GENVIS_API_KEY;
  const baseURL = process.env.GENVIS_BASE_URL || 'https://genvis.xyz/v1';

  if (!apiKey || apiKey === 'sk-your-api-key') {
    throw new Error('Set GENVIS_API_KEY in your environment or .env file.');
  }

  return { apiKey, baseURL };
}

export function createOpenAIClient() {
  return new OpenAI(getConfig());
}

export async function submitJsonVideoTask(payload) {
  const { apiKey, baseURL } = getConfig();
  const response = await fetch(`${baseURL}/video/generations`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  });

  return readJsonResponse(response);
}

export async function getJsonVideoTask(taskId) {
  const { apiKey, baseURL } = getConfig();
  const response = await fetch(`${baseURL}/video/generations/${taskId}`, {
    headers: {
      Authorization: `Bearer ${apiKey}`,
    },
  });

  return readJsonResponse(response);
}

export async function pollVideoTask(taskId, options = {}) {
  const intervalMs = options.intervalMs || 5000;
  const timeoutMs = options.timeoutMs || 10 * 60 * 1000;
  const startedAt = Date.now();

  while (Date.now() - startedAt < timeoutMs) {
    const task = await getJsonVideoTask(taskId);
    const status = task.status || task.data?.status;

    if (status === 'completed' || status === 'succeeded' || status === 'success') {
      return task;
    }

    if (status === 'failed' || status === 'cancelled' || status === 'canceled') {
      throw new Error(`Video task failed: ${JSON.stringify(task)}`);
    }

    await new Promise((resolve) => setTimeout(resolve, intervalMs));
  }

  throw new Error(`Video task timed out after ${timeoutMs} ms: ${taskId}`);
}

async function readJsonResponse(response) {
  const text = await response.text();
  let data;

  try {
    data = text ? JSON.parse(text) : {};
  } catch {
    data = { raw: text };
  }

  if (!response.ok) {
    const message = data.error?.message || data.message || response.statusText;
    throw new Error(`Genvis API error ${response.status}: ${message}`);
  }

  return data;
}
